primitives/src/merkle_tree/field_based_mht/smt/big_lazy_merkle_tree.rs

```diff
@@ -0,0 +1,1038 @@
+        // if root has changed then we cannot get valid merkle path until we finalize it
+        if self.pending_changes() {
+            eprintln!("Identified pending changes: unable to get path before pending changes are applied to the tree.");
+            return None;
+        }
+
+        // check that the index of the leaf is less than the width of the Merkle tree
+        if idx >= self.width {
+            eprintln!("Leaf index out of range. Max: {}, got: {}", self.width - 1, idx);
+            return None;
+        }
+
+        let mut path = Vec::with_capacity(self.height as usize);
+        let mut node_idx = idx;
+        let mut height = 0usize;
+        while height != self.height as usize {
+
+            // Estabilish if sibling is a left or right child
+            let (sibling_idx, direction) = if node_idx % T::MERKLE_ARITY as u32 == 0 {
+                (node_idx + 1, false)
+            } else {
+                (node_idx - 1, true)
+            };
+
+            // Get its hash
+            let (sibling, _) = self.get_node_at_height_and_idx(height, sibling_idx);
+
+            // Push info to path
+            path.push((sibling, direction));
+
+            height += 1; // go up one level
+            node_idx = self.width + (node_idx / T::MERKLE_ARITY as u32); // compute the index of the parent
+        }
+        debug_assert!(self.nodes.get(&node_idx).unwrap() == &self.root.0);
+
+        Some(FieldBasedBinaryMHTPath::<T>::new(path))
+    }
+
+    fn height(&self) -> usize {
+        self.height as usize
+    }
+}
+
+impl<T: BatchFieldBasedMerkleTreeParameters> FieldBasedSparseMerkleTree for FieldBasedOptimizedSparseMHT<T> {
+    fn insert_leaves(&mut self, leaves: Vec<(Self::Position, T::Data)>) -> Result<&mut Self, Error> {
+        self.update_leaves(
+            leaves
+                .into_iter()
+                .map(|(pos, leaf)| OperationLeaf::new(pos, ActionLeaf::Insert, Some(leaf)))
+                .collect()
+        )
+    }
+
+    fn remove_leaves(&mut self, leaves: Vec<Self::Position>) -> Result<&mut Self, Error> {
+        self.update_leaves(
+            leaves
+                .into_iter()
+                .map(|pos| OperationLeaf::new(pos, ActionLeaf::Remove, None))
+                .collect()
+        )
+    }
+
+    /// Perform insertion/removals of the leaves as specified by 'vec_leaf_op'.
+    /// This function will return Error in the following situations:
+    /// - Invalid leaf idx (leaf.coord.position > self.width);
+    /// - Remove a non existing leaf
+    /// Insertion of an already-existing leaf, instead, it's allowed and its value will be simply replaced (as long as it's different from before).
+    /// NOTE: Any duplicated leaf coord in 'vec_leaf_op' will be set to its last (ActionLeaf, T::Data) value.
+    fn update_leaves(&mut self, vec_leaf_op: Vec<OperationLeaf<Self::Position, T::Data>>) -> Result<&mut Self, Error> {
```

Comment:
Done

Reply:
I think we can still keep the Vec here. Since removal is performed by insertion of empty node value, the case in which we insert into the vec a leaf to remove and then insert (or viceversa) should be handled and produces the correct result.